axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/SegmentBasedEventStore.java

```diff
@@ -116,6 +109,54 @@ public SegmentBasedEventStore(EventTypeContext eventTypeContext, IndexManager in
 
     public abstract void handover(Long segment, Runnable callback);
 
+    public Flux<SerializedEvent> eventsPerAggregate(String aggregateId,
+                                                    long firstSequence,
+                                                    long lastSequence,
+                                                    long minToken) {
+        logger.debug("Reading index entries for aggregate {} started.", aggregateId);
+        //Map<segment, all positions in that segment>
+        SortedMap<Long, IndexEntries> positionInfos = indexManager.lookupAggregate(aggregateId,
+                                                                                   firstSequence,
+                                                                                   lastSequence,
+                                                                                   Long.MAX_VALUE,
+                                                                                   minToken);
+        logger.debug("Reading index entries for aggregate {} finished.", aggregateId);
+        List<Flux<SerializedEvent>> fluxList = positionInfos
+                .entrySet()
+                .stream()
+                .map(e -> eventsForPositions(e.getKey(), e.getValue().positions()))
+                .collect(Collectors.toList());
+        final String subscription_time = "SUBSCRIPTION_TIME";
+        return Flux.concat(fluxList)
+                   .doOnEach(signal -> {
+                       if (signal.isOnSubscribe()) {
+                           AtomicLong subscriptionTime = signal.getContextView().get(subscription_time);
+                           subscriptionTime.set(System.currentTimeMillis());
+                       } else if (signal.isOnComplete()) {
+                           AtomicLong subscriptionTime = signal.getContextView().get(subscription_time);
+                           long readTime = System.currentTimeMillis() - subscriptionTime.get();
+                           aggregateReadTimer.record(readTime, TimeUnit.MILLISECONDS);
+                       }
+                   })
+                   .skipUntil(se -> se.getAggregateSequenceNumber() >= firstSequence)
+                   .takeWhile(se -> se.getAggregateSequenceNumber() < lastSequence) //remember: last sequence excluded
+                   .contextWrite(ctx -> ctx.put(subscription_time, new AtomicLong()));
+    }
+
+    private Flux<SerializedEvent> eventsForPositions(long segment, List<Integer> positions) {
+        return (!containsSegment(segment) && next != null) ?
```

Comment:
What if !containsSegment and next == null?

Reply:
In this case, we basically return a Flux that completes immediately as the provider of the EventSource returns an empty Optional.